boring/src/ssl/test/custom_verify.rs

```diff
@@ -0,0 +1,246 @@
+
+            Err(SslVerifyError::Invalid(SslAlert::CERTIFICATE_UNKNOWN))
+        });
+
+    client.connect_err();
+}
+
+#[test]
+fn trusted_with_set_cert() {
+    let server = Server::builder().build();
+    let mut client = server.client_with_root_ca();
+
+    client
+        .ctx()
+        .set_custom_verify_callback(SslVerifyMode::PEER, move |ssl| {
+            let store = ssl.ssl_context().cert_store();
+            let cert = ssl.peer_certificate().unwrap();
+            let cert_chain = ssl.peer_cert_chain().unwrap();
+
+            assert_eq!(store.objects().len(), 1);
+
+            X509StoreContext::new()
+                .unwrap()
+                .init(store, &cert, cert_chain, |store_ctx| {
+                    assert!(store_ctx.verify_cert().unwrap());
+                    assert_eq!(store_ctx.verify_result(), Ok(()));
+
+                    Ok(())
+                })
+                .unwrap();
+
+            Ok(())
+        });
+
+    client.connect();
+}
+
+#[test]
+fn trusted_callback_override_ok() {
+    let server = Server::builder().build();
+    let mut client = server.client_with_root_ca();
+
+    client
+        .ctx()
+        .set_custom_verify_callback(SslVerifyMode::PEER, |ssl| {
+            assert!(ssl.peer_certificate().is_some());
+
+            Ok(())
+        });
+
+    client.connect();
+}
+
+#[test]
+fn trusted_callback_override_bad() {
+    let mut server = Server::builder();
+
+    server.should_error();
+
+    let server = server.build();
+    let mut client = server.client_with_root_ca();
+
+    client
+        .ctx()
+        .set_custom_verify_callback(SslVerifyMode::PEER, |_| {
+            Err(SslVerifyError::Invalid(SslAlert::CERTIFICATE_UNKNOWN))
+        });
+
+    client.connect_err();
+}
+
+#[test]
+fn callback() {
+    static CALLED_BACK: AtomicBool = AtomicBool::new(false);
+    let server = Server::builder().build();
+    let mut client = server.client();
+    let expected = "59172d9313e84459bcff27f967e79e6e9217e584";
+
+    client
+        .ctx()
+        .set_verify_callback(SslVerifyMode::PEER, |_, _| {
+            panic!("verify callback should not be called");
+        });
+
+    client
+        .ctx()
+        .set_custom_verify_callback(SslVerifyMode::PEER, move |ssl| {
+            CALLED_BACK.store(true, Ordering::SeqCst);
+
+            let cert = ssl.peer_certificate().unwrap();
+            let digest = cert.digest(MessageDigest::sha1()).unwrap();
+
+            assert_eq!(hex::encode(digest), expected);
+
+            Ok(())
+        });
+
+    client.connect();
+    assert!(CALLED_BACK.load(Ordering::SeqCst));
+}
+
+#[test]
+fn ssl_callback() {
+    static CALLED_BACK: AtomicBool = AtomicBool::new(false);
+    let server = Server::builder().build();
+    let mut client = server.client().build().builder();
+    let expected = "59172d9313e84459bcff27f967e79e6e9217e584";
+
+    client
+        .ssl()
+        .set_verify_callback(SslVerifyMode::PEER, |_, _| {
```

Comment:
maybe another instance of this test where we set `set_custom_verify_callback` in the SslContext and the Ssl?